File: AMG.Physics/Velocity.cs
//using System;

//namespace AMG.Physics
//{


//    public class Velocity : TimeDependentAction {
//        private readonly IElement _element;

//        public Velocity(IElement element)
//        {
//            _element = element;
//        }

//        public Dimensions Dimensions = new Dimensions(0, 0);

//        public override void Act()
//        {
//            double interval = ResetInterval();
//            _element.Location += Dimensions * interval;
//        }

//        public void Bounce(Dimensions dimensions)
//        {
//            var impulse = (-Dimensions.X * dimensions.X - Dimensions.Y * dimensions.Y);

//            if (impulse > 0.0)
//            {
//                var impulseVector = dimensions * impulse;
//                Dimensions = Dimensions + impulseVector * 2.0;
//            }
//        }
//    }
//} 

File: AMG.FySics/Velocity.fs
namespace AMG.FySics
    type Velocity(Vector : Vector) =        
        member this.Vector = Vector
        member this.Act(location : Vector, interval: float) = 
            location + (Vector * interval)
        member this.Bounce(direction : Unit, loss: double) = 
            let impulse = (-Vector.X * direction.X - Vector.Y * direction.Y)
            let impulseVector = direction * impulse ;
            if impulse > 0.0 then (Vector + (impulseVector * 2.0))* loss else Vector
        member this.Bounce(direction : Unit) = this.Bounce (direction, 1.0)
            
        override this.ToString() = this.Vector.ToString()
